packages/grid/_modules_/grid/hooks/features/pagination/useGridPagination.ts

```diff
@@ -66,60 +56,89 @@ export const useGridPagination = (apiRef: GridApiRef): void => {
 
       setGridState((state) => ({
         ...state,
-        pagination: applyConstraints(
-          {
-            ...state.pagination,
-            pageSize,
-          },
-          options.page,
-        ),
+        pagination: applyConstraints({
+          ...state.pagination,
+          pageSize,
+        }),
       }));
       forceUpdate();
-
-      const params = apiRef.current.getState<GridPaginationState>(
-        'pagination',
-      ) as GridPageChangeParams;
-      apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, {
-        // TODO remove params
-        ...params,
-        pageSize,
-      });
     },
-    [apiRef, setGridState, forceUpdate, logger, options.page],
+    [setGridState, forceUpdate, logger],
   );
 
-  useGridApiOptionHandler(apiRef, GRID_PAGE_CHANGE, options.onPageChange);
-  useGridApiOptionHandler(apiRef, GRID_PAGE_SIZE_CHANGE, options.onPageSizeChange);
+  React.useEffect(() => {
+    apiRef.current.updateControlState<number>({
+      stateId: 'page',
+      propModel: props.page,
+      propOnChange: props.onPageChange,
+      stateSelector: (state) => state.pagination.page,
+      onChangeCallback: (model) => {
+        apiRef.current.publishEvent(GRID_PAGE_CHANGE, model);
+      },
+    });
+  }, [apiRef, props.page, props.onPageChange]);
+
+  React.useEffect(() => {
+    apiRef.current.updateControlState<number>({
+      stateId: 'pageSize',
+      propModel: props.pageSize,
+      propOnChange: props.onPageSizeChange,
+      stateSelector: (state) => state.pagination.pageSize,
+      onChangeCallback: (model) => {
+        apiRef.current.publishEvent(GRID_PAGE_SIZE_CHANGE, model);
+      },
+    });
+  }, [apiRef, props.pageSize, props.onPageSizeChange]);
+
+  React.useEffect(() => {
+    if (options.page === undefined) {
+      return;
+    }
+
+    setGridState((oldState) => ({
+      ...oldState,
+      pagination: {
+        ...oldState.pagination,
+        page: options.page!,
+      },
+    }));
+  }, [setGridState, options.page]);
+
+  React.useEffect(() => {
+    if (
+      options.pageSize === undefined &&
+      (!options.autoPageSize || !containerSizes?.viewportPageSize)
+    ) {
+      return;
+    }
+
+    setGridState((oldState) => ({
+      ...oldState,
+      pagination: {
+        ...oldState.pagination,
+        pageSize: options.autoPageSize ? containerSizes?.viewportPageSize! : options.pageSize!,
```

Comment:
if you set autoPageSize and pageSize then you should be in some kind of invalid state. As if pageSize is controlled or set then, the state won't be updated to the viewportPageSize.

Reply:
I kept the old priority 

```tsx
          pageSize:
            (options.autoPageSize ? containerSizes?.viewportPageSize : options.pageSize) ||
            state.pagination.pageSize,
```

But you're right, the control state should have the priority over the `autoPageSize` to avoid inconsistency.